packages/server/src/db/views/postThread.ts

```diff
@@ -0,0 +1,155 @@
+      })
+      .getRawOne()
+
+    let thread = rowToPost(res)
+    if (depth > 0) {
+      thread.replies = await getReplies(db, thread, depth - 1, requester)
+    }
+
+    return { thread }
+  }
+
+const getReplies = async (
+  db: DataSource,
+  parent: PostThreadView.Post,
+  depth: number,
+  requester: string,
+): Promise<PostThreadView.Post[]> => {
+  const res = await postInfoBuilder(db, requester)
+    .where('post.replyParent = :uri', { uri: parent.uri })
+    .orderBy('post.createdAt', 'DESC')
+    .getRawMany()
+  const got = await Promise.all(
+    res.map(async (row) => {
+      const post = rowToPost(row, parent)
+      if (depth > 0) {
+        post.replies = await getReplies(db, post, depth - 1, requester)
+      }
+      return post
+    }),
+  )
+  return got
+}
+
+// selects all the needed info about a post, just need to supply the `where` clause
+const postInfoBuilder = (db: DataSource, requester: string) => {
+  return db
+    .createQueryBuilder()
+    .select([
+      'post.uri AS uri',
+      'author.did AS authorDid',
+      'author.username AS authorName',
+      'author_profile.displayName AS authorDisplayName',
+      'record.raw AS rawRecord',
+      'reply_count.count AS replyCount',
+      'like_count.count AS likeCount',
+      'repost_count.count AS repostCount',
+      'record.indexedAt AS indexedAt',
+      'requester_repost.uri AS requesterRepost',
+      'requester_like.uri AS requesterLike',
+    ])
+    .from(PostIndex, 'post')
+    .innerJoin(AdxRecord, 'record', 'record.uri = post.uri')
+    .innerJoin(UserDid, 'author', 'author.did = post.creator')
+    .leftJoin(
+      ProfileIndex,
+      'author_profile',
+      'author.did = author_profile.creator',
+    )
+    .leftJoin(
+      util.countSubquery(LikeIndex, 'subject'),
+      'like_count',
+      'like_count.subject = post.uri',
+    )
+    .leftJoin(
+      util.countSubquery(RepostIndex, 'subject'),
+      'repost_count',
+      'repost_count.subject = post.uri',
+    )
+    .leftJoin(
+      util.countSubquery(PostIndex, 'replyParent'),
+      'reply_count',
+      'reply_count.subject = post.uri',
+    )
+    .leftJoin(
+      RepostIndex,
+      'requester_repost',
+      `requester_repost.creator = :requester AND requester_repost.subject = post.uri`,
+      { requester },
+    )
+    .leftJoin(
+      RepostIndex,
+      'requester_like',
+      `requester_like.creator = :requester AND requester_like.subject = post.uri`,
+      { requester },
+    )
+}
+
+// converts the raw SQL output to a Post object
+// unfortunately not type-checked since we're dealing with raw SQL, so change with caution!
+const rowToPost = (
+  row: any,
+  parent?: PostThreadView.Post,
+): PostThreadView.Post => {
+  return {
+    uri: row.uri,
+    author: {
+      did: row.authorDid,
+      name: row.authorName,
+      displayName: row.authorDisplayName || undefined,
+    },
+    record: JSON.parse(row.rawRecord),
+    parent: parent ? { ...parent } : undefined,
```

Comment:
There's a tricky thing here where a `Post` contains references to both the children & the parent which creates a circular reference in the JSON that can't be stringified.

So we clone the parent to break that circular reference.

Eliminating the circular reference means that you can't step into children & then back to the parent without losing information. Not sure the best way to deal with this except for reconstructing the references on the client.

Or maybe we just don't care about the children of a parent & if we need that info we can grab that view.

Not sure if this comment makes sense, lmk if I should clarify 😅

Reply:
Yeah that makes sense. The way I was handling this (IIRC) is that I only set `parent` for ancestors of the target post, and `children` for descendants. Does that make sense?